keanu-project/src/test/java/io/improbable/keanu/plating/LoopTest.java

```diff
@@ -0,0 +1,167 @@
+import io.improbable.keanu.vertices.VertexLabelException;
+import io.improbable.keanu.vertices.VertexMatchers;
+import io.improbable.keanu.vertices.bool.BoolVertex;
+import io.improbable.keanu.vertices.bool.probabilistic.BernoulliVertex;
+import io.improbable.keanu.vertices.dbl.DoubleVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.DoubleProxyVertex;
+
+public class LoopTest {
+
+    @Rule
+    public ExpectedException expectedException = ExpectedException.none();
+
+    private final Function<DoubleVertex, DoubleVertex> increment = (v) -> v.plus(1.);
+    private final Supplier<BoolVertex> flip = () -> new BernoulliVertex(0.5);
+    private final Supplier<BoolVertex> alwaysTrue = () -> ConstantVertex.of(true);
+    private final Vertex startValue = ConstantVertex.of(0.);
+
+    @Test
+    public void youCanGetTheOutputVertex() {
+        Loop loop = Loop
+            .startingFrom(startValue)
+            .whilst(flip)
+            .apply(increment);
+        Vertex<? extends Tensor<?>> output = loop.getOutput();
+        assertThat(output, instanceOf(DoubleVertex.class));
+    }
+
+    @Test
+    public void thereIsADefaultMaxLength(){
+        Loop loop = Loop
+            .startingFrom(startValue)
+            .whilst(alwaysTrue)
+            .apply(increment);
+        assertThat(loop.getPlates().size(), equalTo(Loop.DEFAULT_MAX_COUNT));
+    }
+
+    @Test
+    public void itThrowsIfYouGetTheOutputButTheMaxNumberOfIterationsHasBeenReached() {
+        expectedException.expect(LoopException.class);
+        expectedException.expectMessage("Loop has exceeded its max count");
+        Loop loop = Loop
+            .startingFrom(startValue)
+            .whilst(alwaysTrue)
+            .apply(increment);
+        loop.getOutput();
+    }
+
+    @Test
+    public void youCanTellItNotToThrowWhenTheMaxNumberOfIterationsHaveBeenReached() {
+        Loop loop = Loop
+            .startingFrom(startValue)
+            .dontThrowWhenMaxCountIsReached()
+            .whilst(alwaysTrue)
+            .apply(increment);
+        loop.getOutput();
+        // does not throw
+    }
+
+    @Test
+    public void youCanOverrideTheDefaultMaxLength(){
+        int customMaxCount = 5;
+        Loop loop = Loop
+            .startingFrom(startValue)
+            .atMost(customMaxCount)
+            .whilst(alwaysTrue)
+            .apply(increment);
+        assertThat(loop.getPlates().size(), equalTo(customMaxCount));
+    }
+
+    @Test
+    public void itThrowsIfYouPassInMultipleBaseCaseVertexesAndDontLabelTheOutput(){
+        expectedException.expect(VertexLabelException.class);
+        expectedException.expectMessage("You must pass in a base case, i.e. a vertex labeled as Loop.VALUE_OUT_LABEL");
+        Loop.startingFrom(ConstantVertex.of(0.), ConstantVertex.of(1.))
+            .whilst(alwaysTrue)
+            .apply(increment);
+    }
+
+    @Test
+    public void youCanLoopUntilAConditionIsTrue() {
```

Comment:
Can we add a 2nd loop test? 
```
        Loop loop = Loop
            .startingFrom(startValue)
            .dontThrowWhenMaxCountIsReached()
            .whilst(alwaysTrue)
            .apply(increment);

        Loop loop2 = Loop
            .startingFrom((Vertex<?>) loop.getOutput())
            .dontThrowWhenMaxCountIsReached()
            .whilst(alwaysTrue)
            .apply(increment);

        Vertex<?> output = loop2.getOutput();

        new BayesianNetwork(output.getConnectedGraph());
```

Reply:
That's a really good suggestion because it highlights the fact that VertexLabels are not right for Plates. You need to label the output of `loop` is the input to `loop2`, but then you have 2 vertices with the same label so the BayesNet throws.

Reply:
Done now, after redesigning Plates so that it doesn't rely on the Vertex's label.